akka-docs/rst/scala/code/docs/persistence/PersistenceSchemaEvolutionDocSpec.scala

```diff
@@ -0,0 +1,330 @@
+
+  }
+  
+  case class SeatReserved(code: String, seatType: SeatType)
+  //#protobuf-read-optional-model
+
+  val proto = """
+  //#protobuf-read-optional-proto
+  // FlightAppModels.proto
+  option java_package = "docs.persistence.proto";
+  option optimize_for = SPEED;
+
+  message SeatReserved {
+    required string code     = 1;
+    optional string seatType = 2; // the new field
+  }
+  //#protobuf-read-optional-proto
+  """
+
+  //#protobuf-read-optional
+  /**
+   * Example serializer impl which uses protocol buffers generated classes (proto.*)
+   * to perform the to/from binary marshalling.
+   */
+  class AddedFieldsSerializerWithProtobuf extends Serializer {
+    override def identifier = 67876
+    override def includeManifest = true
+
+    override def fromBinary(bytes: Array[Byte], manifest: Option[Class[_]]): AnyRef =
+      manifest match {
+        case Some(c) if c == classOf[SeatReserved] =>
+          seatReserved(proto.SeatReserved.parseFrom(bytes)) // use generated protobuf serializer
+      }
+
+    override def toBinary(o: AnyRef): Array[Byte] = o match {
+      case s: SeatReserved => 
+        proto.SeatReserved.newBuilder
+          .setCode(s.code)
+          .setSeatType(s.seatType)
+          .build().toByteArray
+    }
+    
+    // -- fromBinary helpers --
+    
+    private def seatReserved(p: proto.SeatReserved): SeatReserved =
+      SeatReserved(p.getCode, seatType(p))
+
+    // handle missing field by assigning "Unknown" value
+    private def seatType(p: proto.SeatReserved): SeatType =
+      if (p.hasSeatType) SeatType.fromString(p.getSeatType) else SeatType.Unknown
+    
+    
+  }
+  //#protobuf-read-optional
+ }
+
+//#simplest-custom-serializer-model
+final case class Person(name: String, surname: String)
+//#simplest-custom-serializer-model
+
+//#simplest-custom-serializer
+/**
+ * Simplest possible serializer, uses a string representation of the Person class.
+ *
+ * Usually a serializer like this would use a library like:
+ * protobuf, kryo, avro, cap'n proto, flatbuffers, SBE or some other dedicated serializer backend
+ * to perform the actual to/from bytes marshalling.
+ */
+class SimplestPossiblePersonSerializer extends Serializer {
+  val Utf8 = Charset.forName("UTF-8")
+
+  // unique identifier of the serializer
+  def identifier = 1234567
+
+  // true if the serializer should include the class manifest automatically (for later deserialization)
+  def includeManifest: Boolean = true
+
+  def toBinary(obj: AnyRef): Array[Byte] = obj match {
+    case p: Person => s"""${p.name}|${p.surname}""".getBytes(Utf8)
+  }
+
+  // since includeManifest was true, we will get the Person clazz here:
+  def fromBinary(bytes: Array[Byte], clazz: Option[Class[_]]): AnyRef = clazz match {
+    case Some(c) if c == classOf[Person] =>
+      val nameAndSurname = new Predef.String(bytes, Utf8)
+      val Array(name, surname) = nameAndSurname.split("[|]")
+      Person(name, surname)
+  }
```

Comment:
This means mismatches will produce nondescript MatchError; should probably raise a more descriptive exception.